quinn-proto/src/congestion.rs

```diff
@@ -1,27 +1,67 @@
 //! Logic for controlling the rate at which data is sent
 
-use std::time::{Duration, Instant};
+use crate::connection::paths::RttEstimator;
+use std::fmt::Debug;
+use std::time::Instant;
 
+mod bbr;
+mod bbr_bw_estimation;
+mod bbr_min_max;
 mod cubic;
 mod new_reno;
 
+pub use bbr::{BBRConfig, BBR};
 pub use cubic::{Cubic, CubicConfig};
 pub use new_reno::{NewReno, NewRenoConfig};
 
 /// Common interface for different congestion controllers
-pub trait Controller: Send {
+pub trait Controller: Send + Debug {
+    /// Packets were sent, called when the send buffer is sent, not when each packet is sent.
+    #[allow(unused_variables)]
+    fn on_sent(&mut self, now: Instant, bytes: u64) {}
+
     /// Packet deliveries were confirmed
     ///
     /// `app_limited` indicates whether the connection was blocked on outgoing
     /// application data prior to receiving these acknowledgements.
-    fn on_ack(&mut self, now: Instant, sent: Instant, bytes: u64, app_limited: bool, rtt: Duration);
+    #[allow(unused_variables)]
+    fn on_ack(
+        &mut self,
+        now: Instant,
+        sent: Instant,
+        bytes: u64,
+        app_limited: bool,
+        rtt: &RttEstimator,
+    ) {
+    }
+
+    /// Packets are acked in batches, all with the same `now` argument. This indicates one of those batches has completed.
+    #[allow(unused_variables)]
+    fn on_end_acks(
+        &mut self,
+        now: Instant,
+        in_flight: u64,
+        app_limited: bool,
+        largest_packet_num_acked: Option<u64>,
+    ) {
+    }
 
     /// Packets were deemed lost or marked congested
     ///
     /// `in_persistent_congestion` indicates whether all packets sent within the persistent
     /// congestion threshold period ending when the most recent packet in this batch was sent were
     /// lost.
-    fn on_congestion_event(&mut self, now: Instant, sent: Instant, is_persistent_congestion: bool);
+    #[allow(unused_variables)]
+    fn on_congestion_event(&mut self, now: Instant, sent: Instant, is_persistent_congestion: bool) {
+    }
+
+    /// Packets were deemed lost. This is only invoked for loss, not congestion signals. `on_congestion_event` is still invoked in the event of loss.
+    #[allow(unused_variables)]
+    fn on_loss(&mut self, now: Instant, sent: Instant, bytes: u64) {}
+
+    /// Updates the last packet number that was sent
+    #[allow(unused_variables)]
+    fn update_last_sent(&mut self, packet_number: u64) {}
```

Comment:
Could this be folded into `on_sent`?

Reply:
collapsed, but please check how it's invoked in quinn-proto/src/connection/mod.rs to make sure it still makes sense